onnxruntime/python/session.py

```diff
@@ -228,7 +228,7 @@ def bind_output(self, name, device_type='cpu', device_id=0, element_type=None, s
         :param shape: output shape
         :param buffer_ptr: memory pointer to output data
         '''
-        if device_type == 'cpu':
+        if buffer_ptr is None:
```

Comment:
What if the buffer_ptr is null and device_type not CPU?

Reply:
Thinking out aloud- I think that will be covered by PR. For example device_type is CUDA and buffer is null, meaning ORT does the allocation (and maintains the life cycle) of the underlying OrtValue (for this release atleast). This will make sure the output is “bound” to CUDA and Run won’t incur the CPU copy cost, but of course to consume the result the user will have to call get_outputs(). This is until we come up with a concrete design (lifecycle management) for some wrapper struct over the ORT allocated device memory the user can take ownership of until they choose to destroy it.

Reply:
If my understanding of pybind is correct (take with a pinch of salt), with the current setup, if the user were to pass None for buffer pointer and with device being non-cpu (cuda), pybind should complain while trying to bind None to the pointer arg in the pybind code

Reply:
It is indeed a problem when buffer_ptr is null and device_type not CPU, which could be covered by Hari's PR. Here are cases for bind_output:
1> buffer_ptr is not null, 
      a) name, shape and element type are all present, covered by current API
      b) otherwise, raise an error
2> buffer_ptr is null, goes to to Hari's API
Make sense, @hariharans29 @pranavsharma?

Reply:
Makes sense. But I think we do not need to raise error if name shape or element type are missing as pybind wont be able to bind the default None as we do not explicitly allow it.

Reply:
I'm inclined to let the users know if the binding failed. I'm going to hold up this PR now. Could you please include the changes we just discussed in your PR?

Reply:
Sure.